app/src/main/java/org/wikipedia/analytics/eventplatform/EventPlatformClient.java

```diff
@@ -0,0 +1,451 @@
+        if (!SamplingController.isInSample(event)) {
+            return;
+        }
+        addEventMetadata(event);
+        OutputBuffer.schedule(event);
+    }
+
+    /**
+     * Supplement the outgoing event with additional metadata, if not already present.
+     * These include:
+     * - client_dt: ISO 8601 timestamp
+     * - app_session_id: the current session ID
+     * - app_install_id: app install ID
+     *
+     * @param event event
+     */
+    static void addEventMetadata(Event event) {
+        if (event.getTimestamp() == null) {
+            event.setTimestamp(getIso8601Timestamp());
+        }
+        event.setSessionId(AssociationController.getSessionId());
+        event.setAppInstallId(getAppInstallId());
+    }
+
+    /**
+     * OutputBuffer: buffers events in a queue prior to transmission
+     *
+     * Transmissions are not sent at a uniform offset but are shaped into
+     * 'bursts' using a combination of queue size and debounce time.
+     *
+     * These concentrate requests (and hence, theoretically, radio awake state)
+     * so as not to contribute to battery drain.
+     */
+    static class OutputBuffer {
+
+        private static List<Event> QUEUE = new ArrayList<>();
+
+        /*
+         * When an item is added to QUEUE, wait this many ms before sending.
+         *
+         * If another item is added to QUEUE during this time, reset the
+         * countdown.
+         */
+        private static final int WAIT_MS = 30000;
+
+        /*
+         * When QUEUE.size() exceeds this value TIMER becomes non-interruptable.
+         */
+        private static final int MAX_QUEUE_SIZE = 128;
+
+        /*
+         * IMPLEMENTATION NOTE: Java Timer will provide the desired asynchronous
+         * countdown after a new item is added to QUEUE.
+         */
+        private static Timer TIMER = new Timer();
+
+        /*
+         * IMPLEMENTATION NOTE: Java abstract TimerTask class requires a run()
+         * method be defined.
+         *
+         * The run() method is called when the Timer expires.
+         */
+        private static class Task extends TimerTask {
+            public void run() {
+                sendAllScheduled();
+            }
+        }
+
+        /**
+         * If sending is enabled, dequeue and call send() on all scheduled items.
+         */
+        static synchronized void sendAllScheduled() {
+            TIMER.cancel();
+
+            if (ENABLED) {
+                /*
+                 * All items on QUEUE are permanently removed.
+                 */
+                send(QUEUE);
+                QUEUE.clear();
+            }
+        }
+
+        /**
+         * Schedule a request to be sent.
+         *
+         * @param event event data
+         */
+        static synchronized void schedule(Event event) {
+            /*
+             * Item is enqueued whether or not sending is enabled.
+             */
+            QUEUE.add(event);
```

Comment:
If I'm understanding correctly, if `ENABLED` is false, then the `QUEUE` will continue to grow without bounds, and will eventually run out of memory. There should probably be an upper limit on the number of items in the queue.